api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,154 @@
+ */
+enum class WasmOnVmStartResult : uint32_t {
+  Ok = 0,
+  BadConfiguration = 1,
+};
+extern "C" WasmOnVmStartResult proxy_on_vm_start(uint32_t root_context_id,
+                                                 uint32_t vm_configuration_size);
+
+/**
+ * Can be called to validate a configuration (e.g. from bootstrap or xDS) both before
+ * proxy_on_start() to verify the VM configuration or after proxy_on_start() to verify a plugin
+ * configuration.
+ * @param root_context_id is a unique identifier for the configuration verification context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (i.e. bad configuration).
+ */
+enum class WasmOnValidateConfigurationResult : uint32_t {
+  Ok = 0,
+  BadConfiguration = 1,
+};
+extern "C" WasmOnValidateConfigurationResult
+proxy_validate_configuration(uint32_t root_context_id, uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
+ * @param root_context_id is an identifier for one or more related plugins.
+ * @param plugin_configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+enum class WasmOnConfigureResult : uint32_t {
+  Ok = 0,
+  BadConfiguration = 1,
+}
+extern "C" WasmOnConfigureResult proxy_on_configure(uint32_t root_context_id,
+                                                uint32_t plugin_configuration_size);
+
+// Stream calls.
+
+/**
+ * Called when a request, stream or other ephemeral context is created.
+ * @param context_id is an identifier of the ephemeral context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ */
+extern "C" void proxy_on_context_create(uint32_t context_id, uint32_t root_context_id);
+
+// Stream and Non-stream calls.
+
+/**
+ * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
+ * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
+ * called when the VM is going to shutdown.
+ * @param context_id is an identifier the context.
+ * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
+ * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
+ * future proxy_done() call by the root context.
+ */
+enum class WasmOnDoneResult : uint32_t {
+  Done = 0,
+  NotDone = 1,
+}
+extern "C" WasmOnDoneResult proxy_on_done(uint32_t context_id);
+
+/**
+ * Called when the context is being deleted and will no longer receive any more calls.
+ * @param context_id is an identifier the context.
+ */
+extern "C" void proxy_on_delete(uint32_t context_id);
+
+//
+// ABI calls from the VM to the host.
```

Comment:
One last nit: do we want to potentially have a different naming scheme for calls from host --> VM and VM --> host to make it clear the directionality here? E.g. `proxy_vm_foo`, `proxy_host__bar`?

Reply:
All host -> VM functions ("callbacks") are named `proxy_on_<something>`
All VM -> host functions are named `proxy_<verb>_<something>`.

Reply:
@PiotrSikora yes, it's technically correct (the best kind) if you stare at it, I was asking for something a bit more explicit. I will defer to what @lizan and @mattklein123 feel here, as they are also new to the code base and might find `proxy_on` clear enough or prefer making it more explicit.

Reply:
In a perfect world I agree a more obvious convention here would be useful.

Reply:
I am open to suggestions.

Reply:
It's not great, but something like:
`proxy_vth_*` (VM to host)
`proxy_htv_*` (host to VM)

?

Reply:
That is very ugly.   I think proxy_on_ makes sense as the "on" prefix is widely used for hooks, e.g. in HTML/javascript.

Reply:
proxy_vth_ is also not readily understandable.  If I saw that I would just have to google it which kind of defeats the purpose of a naming convention.

Reply:
I'm fine with keeping it as is. The other option is to spell it out which isn't that long: `proxy_vm_to_host_*` and `proxy_host_to_vm_*` ?

Reply:
From Envoy point of view, callbacks (`proxy_on_http_request_headers`, `proxy_on_http_response_body`, etc.) are just a glue code extending existing callbacks (`decodeHeaders`, `encodeData`, etc.), and host functions are everything else.

From WASM module developer point of view, they need to expose a few callbacks (`proxy_on_http_request_headers`, `proxy_on_http_response_body`, etc.), and when they want to get something from host, they just call a function (`proxy_get_http_request_headers`, `proxy_send_http_response`, etc.).

Those two world views don't mix, so there is no place for confusion.

Also, everything (including NullVM extensions) is going to be written using C++ SDK, and not the low-level WASM ABI, so this is completely irrelevant to anybody but WASM integrators.

IMHO, you're are trying to solve a problem that doesn't exist. Please leave it as-is.

Reply:
That's fine. Please update the file location and we can merge and move on.

/wait

Reply:
OK, moved to include/abi/wasm PTAL

Reply:
Can we at least put them in distinct header files? If the two worlds never meet, they should have no need to be in the same header.

Reply:
Done.  PTAL